File: weight_tracker.py

#!/usr/bin/env python3
# weight_tracker.py
import argparse
import json
import csv
import os
import sys
from datetime import datetime, timedelta
from colorama import init, Fore, Style

init(autoreset=True)

DATA_FILE = "weight.json"

class WeightTracker:
    def __init__(self):
        self.records = []
        self.load()

    def load(self):
        if not os.path.exists(DATA_FILE):
            return
        try:
            with open(DATA_FILE, 'r') as f:
                self.records = json.load(f)
        except:
            self.records = []

    def save(self):
        with open(DATA_FILE, 'w') as f:
            json.dump(self.records, f, indent=2)

    def add_record(self, weight, date=None, photo=None):
        if date is None:
            date = datetime.now().strftime("%Y-%m-%d")
        record = {"date": date, "weight": weight}
        if photo:
            record["photo"] = photo
        self.records.append(record)
        self.records.sort(key=lambda x: x["date"])
        self.save()
        print(Fore.GREEN + f"Запись добавлена: {date} - {weight} кг" + (f" (фото: {photo})" if photo else ""))

    def list_records(self):
        if not self.records:
            print(Fore.YELLOW + "Нет записей.")
            return
        print(Fore.CYAN + "Дата       | Вес (кг) | Фото")
        for r in self.records:
            photo = r.get("photo", "-")
            print(f"{r['date']} | {r['weight']:.1f} | {photo}")

    def show_chart(self):
        if not self.records:
            print(Fore.YELLOW + "Нет данных для графика.")
            return
        weights = [r["weight"] for r in self.records]
        min_w = min(weights)
        max_w = max(weights)
        if max_w == min_w:
            print("Все значения одинаковы.")
            return
        scale = 20 / (max_w - min_w)
        print(Fore.CYAN + "График изменения веса (кг):")
        for r in self.records:
            bar_len = int((r["weight"] - min_w) * scale) + 1
            bar = "█" * bar_len
            photo = r.get("photo", "")
            print(f"{r['date']} {Fore.GREEN}{bar} {r['weight']:.1f}  {photo}")

    def show_stats(self):
        if not self.records:
            print(Fore.YELLOW + "Нет данных.")
            return
        weights = [r["weight"] for r in self.records]
        min_w = min(weights)
        max_w = max(weights)
        avg_w = sum(weights) / len(weights)
        first = weights[0]
        last = weights[-1]
        diff = last - first
        trend = "📈 растёт" if diff > 0 else "📉 падает" if diff < 0 else "➡️ стабилен"
        print(Fore.CYAN + "Статистика:")
        print(f"  Минимальный: {Fore.GREEN}{min_w:.1f} кг")
        print(f"  Максимальный: {Fore.RED}{max_w:.1f} кг")
        print(f"  Средний: {Fore.YELLOW}{avg_w:.1f} кг")
        print(f"  Тренд: {Fore.MAGENTA}{trend} ({diff:+.1f} кг)")

    def export(self, filename):
        ext = os.path.splitext(filename)[1].lower()
        if ext == '.json':
            with open(filename, 'w') as f:
                json.dump(self.records, f, indent=2)
        elif ext == '.csv':
            with open(filename, 'w', newline='') as f:
                writer = csv.writer(f)
                writer.writerow(["date", "weight", "photo"])
                for r in self.records:
                    writer.writerow([r["date"], r["weight"], r.get("photo", "")])
        else:
            print(Fore.RED + "Неподдерживаемый формат. Используйте .json или .csv")
            return
        print(Fore.GREEN + f"Экспортировано в {filename}")

def main():
    parser = argparse.ArgumentParser(description="Трекер веса с фотографиями")
    parser.add_argument("--add", type=float, help="Добавить вес (кг)")
    parser.add_argument("--date", help="Дата (YYYY-MM-DD)")
    parser.add_argument("--photo", help="Путь к фотографии")
    parser.add_argument("--list", action="store_true", help="Показать записи")
    parser.add_argument("--chart", action="store_true", help="Показать график")
    parser.add_argument("--stats", action="store_true", help="Показать статистику")
    parser.add_argument("--export", help="Экспорт в файл (JSON/CSV)")
    args = parser.parse_args()

    tracker = WeightTracker()
    if args.add is not None:
        tracker.add_record(args.add, args.date, args.photo)
    elif args.list:
        tracker.list_records()
    elif args.chart:
        tracker.show_chart()
    elif args.stats:
        tracker.show_stats()
    elif args.export:
        tracker.export(args.export)
    else:
        parser.print_help()

if __name__ == "__main__":
    main()

File: weight_tracker.cpp
// weight_tracker.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <json/json.h> // using jsoncpp

using namespace std;

const string DATA_FILE = "weight.json";

struct Record {
    string date;
    double weight;
    string photo;
};

class Tracker {
private:
    vector<Record> records;

    void load() {
        ifstream ifs(DATA_FILE);
        if (!ifs) return;
        Json::Value root;
        ifs >> root;
        for (const auto& item : root) {
            Record r;
            r.date = item["date"].asString();
            r.weight = item["weight"].asDouble();
            r.photo = item.get("photo", "").asString();
            records.push_back(r);
        }
        sort(records.begin(), records.end(), [](const Record& a, const Record& b) {
            return a.date < b.date;
        });
    }

    void save() {
        Json::Value root(Json::arrayValue);
        for (const auto& r : records) {
            Json::Value item;
            item["date"] = r.date;
            item["weight"] = r.weight;
            if (!r.photo.empty()) item["photo"] = r.photo;
            root.append(item);
        }
        ofstream ofs(DATA_FILE);
        ofs << root.toStyledString();
    }

    string today() {
        time_t t = time(nullptr);
        tm* now = localtime(&t);
        char buf[11];
        strftime(buf, sizeof(buf), "%Y-%m-%d", now);
        return string(buf);
    }

public:
    Tracker() { load(); }

    void addRecord(double weight, const string& date, const string& photo) {
        string d = date.empty() ? today() : date;
        records.push_back({d, weight, photo});
        sort(records.begin(), records.end(), [](const Record& a, const Record& b) {
            return a.date < b.date;
        });
        save();
        cout << "\033[32mЗапись добавлена: " << d << " - " << weight << " кг";
        if (!photo.empty()) cout << " (фото: " << photo << ")";
        cout << "\033[0m" << endl;
    }

    void listRecords() {
        if (records.empty()) {
            cout << "\033[33mНет записей.\033[0m" << endl;
            return;
        }
        cout << "\033[36mДата       | Вес (кг) | Фото\033[0m" << endl;
        for (const auto& r : records) {
            string photo = r.photo.empty() ? "-" : r.photo;
            cout << r.date << " | " << fixed << setprecision(1) << r.weight << " | " << photo << endl;
        }
    }

    void showChart() {
        if (records.empty()) {
            cout << "\033[33mНет данных для графика.\033[0m" << endl;
            return;
        }
        double minW = records[0].weight, maxW = records[0].weight;
        for (const auto& r : records) {
            if (r.weight < minW) minW = r.weight;
            if (r.weight > maxW) maxW = r.weight;
        }
        if (maxW == minW) {
            cout << "Все значения одинаковы." << endl;
            return;
        }
        double scale = 20.0 / (maxW - minW);
        cout << "\033[36mГрафик изменения веса (кг):\033[0m" << endl;
        for (const auto& r : records) {
            int barLen = (int)((r.weight - minW) * scale) + 1;
            string bar(barLen, '█');
            string photo = r.photo.empty() ? "" : " " + r.photo;
            cout << r.date << " \033[32m" << bar << "\033[0m " << fixed << setprecision(1) << r.weight << photo << endl;
        }
    }

    void showStats() {
        if (records.empty()) {
            cout << "\033[33mНет данных.\033[0m" << endl;
            return;
        }
        double minW = records[0].weight, maxW = records[0].weight, sum = 0;
        for (const auto& r : records) {
            if (r.weight < minW) minW = r.weight;
            if (r.weight > maxW) maxW = r.weight;
            sum += r.weight;
        }
        double avgW = sum / records.size();
        double first = records.front().weight;
        double last = records.back().weight;
        double diff = last - first;
        string trend = diff > 0 ? "📈 растёт" : diff < 0 ? "📉 падает" : "➡️ стабилен";
        cout << "\033[36mСтатистика:\033[0m" << endl;
        cout << "  Минимальный: \033[32m" << fixed << setprecision(1) << minW << "\033[0m кг" << endl;
        cout << "  Максимальный: \033[31m" << maxW << "\033[0m кг" << endl;
        cout << "  Средний: \033[33m" << avgW << "\033[0m кг" << endl;
        cout << "  Тренд: \033[35m" << trend << "\033[0m (" << showpos << diff << " кг)" << endl;
    }

    void exportData(const string& filename) {
        string ext = filename.substr(filename.find_last_of('.') + 1);
        if (ext == "json") {
            ofstream ofs(filename);
            Json::Value root(Json::arrayValue);
            for (const auto& r : records) {
                Json::Value item;
                item["date"] = r.date;
                item["weight"] = r.weight;
                if (!r.photo.empty()) item["photo"] = r.photo;
                root.append(item);
            }
            ofs << root.toStyledString();
        } else if (ext == "csv") {
            ofstream ofs(filename);
            ofs << "date,weight,photo\n";
            for (const auto& r : records) {
                ofs << r.date << "," << fixed << setprecision(1) << r.weight << "," << r.photo << "\n";
            }
        } else {
            cerr << "\033[31mНеподдерживаемый формат. Используйте .json или .csv\033[0m" << endl;
            return;
        }
        cout << "\033[32mЭкспортировано в " << filename << "\033[0m" << endl;
    }
};

int main(int argc, char* argv[]) {
    double add = 0;
    string date, photo, exportFile;
    bool list = false, chart = false, stats = false;

    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "--add" && i+1 < argc) add = stod(argv[++i]);
        else if (arg == "--date" && i+1 < argc) date = argv[++i];
        else if (arg == "--photo" && i+1 < argc) photo = argv[++i];
        else if (arg == "--list") list = true;
        else if (arg == "--chart") chart = true;
        else if (arg == "--stats") stats = true;
        else if (arg == "--export" && i+1 < argc) exportFile = argv[++i];
    }

    Tracker tracker;
    if (add != 0) {
        tracker.addRecord(add, date, photo);
    } else if (list) {
        tracker.listRecords();
    } else if (chart) {
        tracker.showChart();
    } else if (stats) {
        tracker.showStats();
    } else if (!exportFile.empty()) {
        tracker.exportData(exportFile);
    } else {
        cout << "Используйте --help для справки." << endl;
    }
    return 0;
}
